WordPress/gutenberg
# Components: Add onFocusLoss option to withFocusReturn

Fixes #1918
Addresses https://github.com/WordPress/gutenberg/pull/10215#issuecomment-429959235

This pull request seeks to enhance `withFocusReturn` to handle cases of focus loss by accepting an `onFocusLoss` handler. This handler can be passed as an option when calling `withFocusReturn`, or applied as fallback on a virtual hierarchy using the new `wp.components.FocusReturnProvider` context component. Furthermore, the `Modal` component accepts the `onFocusLoss` prop to allow customized handling.

To this end, the following behaviors take place:

- When a modal opened through "More Menu" options is closed, the "More Menu" button is focused
- When the sidebar is closed, it will direct focus to the Settings button if focus loss were to otherwise occur
- All other focus loss (if exists, not intended) would redirect focus to the `edit-post-header` wrapper

In all cases, the specific behavior can be customized if there's a more suitable alternative presented.

**Pending tasks:**

This could do for some improved unit and/or end-to-end tests, pending feedback on direction.

**Testing instructions:**

Repeat Steps to Reproduce from #1918 and https://github.com/WordPress/gutenberg/pull/10215#issuecomment-429959235, verifying that focus loss does not occur.
